File: scripts/fetch_attendance.py
import requests, json, re, sys
from datetime import datetime, timezone, timedelta
from bs4 import BeautifulSoup

JST = timezone(timedelta(hours=9))
now = datetime.now(JST)
today_str = f"{now.month}/{now.day}"

URL = 'https://www.cityheaven.net/fukuoka/A4001/A400102/royallips21/attend/?shopmenu=2&lo=1'
HEADERS = {
    'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
    'Accept-Language': 'ja-JP,ja;q=0.9,en;q=0.8',
    'Referer': 'https://www.cityheaven.net/fukuoka/A4001/A400102/royallips21/',
}

try:
    r = requests.get(URL, headers=HEADERS, timeout=30)
    r.raise_for_status()
    print(f"Fetched OK ({len(r.text)} bytes)")
except Exception as e:
    print(f"Fetch failed: {e}", file=sys.stderr)
    sys.exit(1)

soup = BeautifulSoup(r.text, 'html.parser')
girls = []
parsed_today = False

# Strategy 1: find today's column in the weekly schedule table
for table in soup.find_all('table'):
    header_row = table.find('tr')
    if not header_row:
        continue
    header_cells = header_row.find_all(['th', 'td'])
    today_col = next(
        (i for i, th in enumerate(header_cells) if today_str in th.get_text()),
        None
    )
    if today_col is None:
        continue
    for row in table.find_all('tr')[1:]:
        cells = row.find_all(['td', 'th'])
        if len(cells) <= today_col:
            continue
        cell_text = cells[today_col].get_text(strip=True)
        if not cell_text or cell_text in ('-', '　', '休'):
            continue
        m = re.search(r'girlid-(\d+)', str(row))
        if m:
            girls.append(m.group(1))
    parsed_today = True
    print(f"Parsed today's column (col={today_col}, date={today_str})")
    break

# Strategy 2: fallback — all girlids on page
if not girls:
    print("No today column found, using all girlids on page")
    girls = list(set(re.findall(r'girlid-(\d+)', r.text)))

girls = list(set(girls))
data = {
    'girlidList': girls,
    'updated': now.isoformat(),
    'date': today_str,
    'parsedToday': parsed_today,
    'count': len(girls),
}
print(f"Result: {len(girls)} girls → {girls}")
with open('attend.json', 'w', encoding='utf-8') as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

File: scripts/fetch_attendance.js
const fs = require('fs');

const JST = 9 * 60 * 60 * 1000;
const now = new Date(Date.now() + JST);
const month = now.getUTCMonth() + 1;
const day = now.getUTCDate();
const todayStr = `${month}/${day}`;
const STORE = 'https://www.cityheaven.net/fukuoka/A4001/A400102/royallips21';
const CONCURRENCY = 8;

function extractText(html) {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, '')
    .replace(/<style[\s\S]*?<\/style>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>')
    .replace(/&nbsp;/g, ' ').replace(/\s+/g, ' ');
}

async function fetchCastInfo(gid) {
  const url = `${STORE}/girlid-${gid}/`;
  try {
    const res = await fetch(url, {
      headers: {
        'Accept-Language': 'ja-JP,ja;q=0.9',
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      },
      signal: AbortSignal.timeout(15000),
    });
    if (!res.ok) { console.log(`${gid}: HTTP ${res.status}`); return null; }

    const text = extractText(await res.text());

    // 今日の確定出勤を探す: "6/29(月) 8:00～14:00" のような形式
    const dateRe = new RegExp(`${month}/${day}[^\\d〜～]{0,10}([\\d]{1,2}:[\\d]{2})[〜～～]([\\d]{1,2}:[\\d]{2})`);
    const scheduleMatch = text.match(dateRe);
    if (!scheduleMatch) return null;

    // 最短案内時刻: "最短XX:XX" のような形式（出勤中のみ表示）
    const earliestMatch = text.match(/最短[^\\d]{0,6}([\\d]{1,2}:[\\d]{2})/);

    return {
      shiftStart: scheduleMatch[1],
      shiftEnd: scheduleMatch[2],
      shiftTime: `${scheduleMatch[1]}〜${scheduleMatch[2]}`,
      earliest: earliestMatch ? earliestMatch[1] : null,
    };
  } catch (e) {
    console.error(`${gid}: ${e.message}`);
    return null;
  }
}

(async () => {
  const casts = JSON.parse(fs.readFileSync('casts.json', 'utf8'));
  const gids = [...new Set(casts.map(c => c.gid).filter(Boolean))];
  console.log(`Checking ${gids.length} casts for ${todayStr}...`);

  const results = new Map();
  for (let i = 0; i < gids.length; i += CONCURRENCY) {
    const batch = gids.slice(i, i + CONCURRENCY);
    const batchResults = await Promise.all(batch.map(fetchCastInfo));
    batch.forEach((gid, j) => { if (batchResults[j]) results.set(gid, batchResults[j]); });
    console.log(`Processed ${Math.min(i + CONCURRENCY, gids.length)}/${gids.length}`);
  }

  const working = [...results.entries()];
  console.log(`Working today (${todayStr}): ${working.length} casts`);
  working.forEach(([gid, info]) =>
    console.log(`  ${gid}: ${info.shiftTime}${info.earliest ? ` 最短${info.earliest}` : ''}`)
  );

  const attendInfo = {};
  working.forEach(([gid, info]) => { attendInfo[gid] = info; });

  fs.writeFileSync('attend.json', JSON.stringify({
    girlidList: working.map(([gid]) => gid),
    attendInfo,
    updated: new Date(Date.now() + JST).toISOString(),
    date: todayStr,
    count: working.length,
  }, null, 2));
})();
